src/ray/core_worker/actor_manager.cc

```diff
@@ -91,6 +91,11 @@ std::pair<std::shared_ptr<const ActorHandle>, Status> ActorManager::GetNamedActo
       RAY_LOG(ERROR) << error_str;
       return std::make_pair(nullptr, Status::TimedOut(error_str));
     }
+  } else {
+    // When the named actor is already cached, the reference of actor_creation_return_id
```

Comment:
NOTE:
Before, the `owner` of actor will always cache miss, so that the `ray.get_actor(actor_name)` will always get actor from GCS, and then call `AddNewActorHandle` to ensure the rcorrectness of actor's reference.

This PR: the `owner` of actor can still get named actor from `cached_actor_name_to_ids_`, so we need add this logic to avoid bad actor's reference.

Why the `owner` of actor will always cache miss?
1. `CoreWorker::CreateActor` --> `ActorManager::AddNewActorHandle` --> `ActorManager::AddNewActorHandle(empty_actor_name)` --> `ActorManager::AddActorHandle` will insert a new actor handler to `actor_handles_`, and the empty_actor_name will never be put into `cached_actor_name_to_ids_`.
 2. Then, `ray.get_actor(actor_name)` will get named actor from GCS, the chain is `CoreWorker::GetNamedActorHandle` --> `ActorManager::GetNamedActorHandle` --> `Actors().SyncGetByName` --> `ActorManager::AddNewActorHandle`. The subcripiton will not be invoked anymore due to there is already an actor handle inside `actor_handles_` generated in step1.

So, no mater how many times the `ray.get_actor(actor_name)` is invoked, the named actor will never be put into `cached_actor_name_to_ids_`.

cc @rkooo567 @stephanie-wang

Reply:
It's better to AddNewActorHandle here?

auto cloned_actor_handle = ...
cached_actor_name = ...
AddNewActorHandle(std::move(cloned_actor_handle),
                  cached_actor_name,
                  call_site,
                  caller_address,
                  /*is_detached*/ true);

Reply:
Also why can't we just update `cached_actor_name_to_ids_ ` when we call `AddOwnedObject` in `AddNewActorHandle`?

Reply:
> cached_actor_name_to_ids_

Actually, I'm not sure if there's a problem with doing what you say, and I think this PR does not involve the change of this logic :)

Reply:
> > So, no mater how many times the ray.get_actor(actor_name) is invoked, the named actor will never be put into cached_actor_name_to_ids_.
> 
> This is the case "after" this PR right? Previously, it was updated because Add new actor handle will subscribe an actor which will update `cached_actor_name_to_ids_ `

I mean, "before" this PR, there is a bug that the cache will never be hit on the owner side, because the `cached_actor_name` is empty.

```
bool ActorManager::AddNewActorHandle(std::unique_ptr<ActorHandle> actor_handle,
                                     const std::string &call_site,
                                     const rpc::Address &caller_address,
                                     bool is_detached) {
  return AddNewActorHandle(std::move(actor_handle),
                           /*cached_actor_name=*/"",
                           call_site,
                           caller_address,
                           is_detached);
}
```

Reply:
> Yes, "before" this PR, there is a bug that the cache will never be hit on the owner side, because the cached_actor_name is empty.

So what I am confused here is that you said you fix the cache miss "bug", but it introduces a new bug that you described here 
```
The named actor is destroyed but the caller does not receives the notification yet, and the execption ("The actor is dead because because all references to the actor were removed") will be raised if it invoke the ray.get(actor_name).
```

What I am confused is -> did the same issue exist "before you fix the cache miss issue"? I am asking it because cache miss is performance optimization whereas the bug you mentioned seems to be a correctness issue. If we have to sacrifice correctness for the performance, I think the fix is incorrect, and we need to discuss a different solution.

Reply:
Thanks for your explanation.

First of all, I think since cache is introduced, there is no reason for it to be invalidated in owner all the time, so I consider it a bug rather than a feature.

Secondly, this PR breaks the previous rule of owner side cache invalidation, which will expose some  new problems. However, these exposed problems will be resolved in this PR, so there will be no correctness problems.

Reply:
Gotcha. Thanks for the clarification. To make sure, you are saying 

1. cache issue is fixed from this PR (owner shouldn't cache miss)
2. And this introduces a new issue
3. But that's also resolved from this PR.

And how you resolved 3 is to add a new reference whenever cache is hit from GetNamedActor right?

Reply:
Yes, it is.

Before the cache was introduced, every time the named actor was obtained from GCS, "AddNewActorHandle" was called to increase the reference count. So, I think there is no reason to skip the `AddNewActorHandle or AddActorHandle` when the named actor was obtained from the cache.

Reply:
BTW, the following test in master branch will get a warning log "Tried to decrease ref count for nonexistent object ID: ...".

I found that when the actor handle reference count is 0, the worker (not owner) can still submit tasks with it. I'm not sure whether it meets the expectation, so I did not add the following test to this PR.

```
def test_named_actor_reference_count(ray_start_regular_shared):
    """
    Test keep calling get_actor.
    """

    @ray.remote
    class Actor:
        def hello(self):
            return "hello"

    @ray.remote
    class ProxyActor:
        def __init__(self):
            self.actor_handle = None

        def hello(self):
            self.actor_handle = ray.get_actor("ABC")
            return ray.get(self.actor_handle.hello.remote())


    actor = Actor.options(name="ABC").remote()
    assert ray.get(actor.hello.remote()) == "hello"

    proxy_actor = ProxyActor.remote()
    # The actor handle will hold by proxy_actor after the next line.
    assert ray.get(proxy_actor.hello.remote()) == "hello"

    # The reference count of actor_handle inside ProxyActor will be 0 after the next line (Which is unexpected).
    assert ray.get(proxy_actor.hello.remote()) == "hello"

    # The warning log "Tried to decrease ref count for nonexistent object ID: ..." will be print after the next line.
    assert ray.get(proxy_actor.hello.remote()) == "hello"
```

cc @rkooo567

Reply:
Nice! Seems like the PR also fixes a hidden bug, so it's great!